# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediacontrol/MediaStatusManager.h
# Commit: 714b2fb21e23
# Full Hash: 714b2fb21e2341e8e2bf1be1271ca0ed5e526443
# Author: alwu <alwu@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1640998
# File Overlap Count: 1
# Description:
#   Bug 1640998 - part5 : set supported keys to the event source r=chunmin
#   
#   This patch will
#   - add a method `SetSupportedMediaKeys()` on `MediaControlKeysEventSource`
#   - set main controller's supported key to the event source
# ==============================================================================

diff -r a48b6554ba9c -r 714b2fb21e23 dom/media/mediacontrol/MediaStatusManager.h
--- a/dom/media/mediacontrol/MediaStatusManager.h	Mon Jun 08 22:16:46 2020 +0000
+++ b/dom/media/mediacontrol/MediaStatusManager.h	Tue Jun 09 02:23:35 2020 +0000
@@ -176,6 +176,13 @@
   ~MediaStatusManager() = default;
   virtual void HandleActualPlaybackStateChanged() = 0;
 
+  // This event would be notified when the active media session changes its
+  // supported actions.
+  MediaEventSource<nsTArray<MediaSessionAction>>&
+  SupportedActionsChangedEvent() {
+    return mSupportedActionsChangedEvent;
+  }
+
   uint64_t mTopLevelBrowsingContextId;
 
   // Within a tab, the Id of the browsing context which has already created a
@@ -194,6 +201,13 @@
   void ClearActiveMediaSessionContextIdIfNeeded();
   void HandleAudioFocusOwnerChanged(Maybe<uint64_t>& aBrowsingContextId);
 
+  void NotifySupportedKeysChangedIfNeeded(uint64_t aBrowsingContextId);
+
+  // Return a copyable array filled with the supported media session actions.
+  // Use copyable array so that we can use the result as a parameter for the
+  // media event.
+  CopyableTArray<MediaSessionAction> GetSupportedActions() const;
+
   // When the amount of playing media changes, we would use this function to
   // update the guessed playback state.
   void SetGuessedPlayState(MediaSessionPlaybackState aState);
@@ -225,6 +239,8 @@
 
   nsDataHashtable<nsUint64HashKey, MediaSessionInfo> mMediaSessionInfoMap;
   MediaEventProducer<MediaMetadataBase> mMetadataChangedEvent;
+  MediaEventProducer<nsTArray<MediaSessionAction>>
+      mSupportedActionsChangedEvent;
   MediaPlaybackStatus mPlaybackStatusDelegate;
 };
 